### PR TITLE
Initial skeleton for Evaluator classes and exceptions

### PR DESCRIPTION
**Pull request by nathan-weinberg**
This PR introduces the `Evaluator` parent class and the `MMLUEvaluator` and `MTBenchEvaluator` child classes

It also introduces the `EvalError` parent exception class for all `eval` custom exceptions and the `ModelNotFoundError` child exception class

### CODE REVIEW

**Review by alinaryan — COMMENTED** [review]
This LGTM - are PR-MMLU and PR-Bench going to live in the MMLU and MT-Bench child classes respectively or is the idea to add them in later as their own child classes?

**Comment by nathan-weinberg** [comment]
> This LGTM - are PR-MMLU and PR-Bench going to live in the MMLU and MT-Bench child classes respectively or is the idea to add them in later as their own child classes?

Good question! I think we could go with either approach - either extending the current child classes or adding new child classes but in the same files.

Could you do an explicit approval? I've set the repo so two are required for merge. @alimaredia same for you.

@danmcp and @xukai92 feel free to weigh-in as well - this foundation will define subsequent work.

**Review by alinaryan — APPROVED** [review]
I lean towards new child classes since the inputs/dependencies differ across PR benchmarks. You could add them in this PR or it can be handled in a follow-up PR by anyone. 
Thanks for working on this!!

**Comment by danmcp** [comment]
> This LGTM - are PR-MMLU and PR-Bench going to live in the MMLU and MT-Bench child classes respectively or is the idea to add them in later as their own child classes?
> @danmcp and @xukai92 feel free to weigh-in as well - this foundation will define subsequent work.

It would be good to understand the signatures before deciding.  But I'm pretty sure they should be new classes.  The other thing to consider whether the modeling makes more sense as a is-a or has-a relationship.  Has-a might give some useful flexibility.

**Comment by nathan-weinberg** [comment]
> > This LGTM - are PR-MMLU and PR-Bench going to live in the MMLU and MT-Bench child classes respectively or is the idea to add them in later as their own child classes?
> > @danmcp and @xukai92 feel free to weigh-in as well - this foundation will define subsequent work.
> 
> It would be good to understand the signatures before deciding. But I'm pretty sure they should be new classes. The other thing to consider whether the modeling makes more sense as a is-a or has-a relationship. Has-a might give some useful flexibility.

I pushed up two additional child classes based on the specs in the doc - lmk your thoughts!

File: src/instructlab/eval/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+class Evaluator:
+    """
+    Parent class for Evaluators
+    """
+
+    def __init__(self, model) -> None:
+        self.model = model
+
+    def run(self) -> dict:
```
[the remark this answers is not in the document]
Comment (nathan-weinberg) [reply src/instructlab/eval/evaluator.py:12]: My assumption here was that the generic `run` function would basically take the class attributes and return a single dict for the CLI to consume - thus the following assumption that all Evaluators would return a dict, though with different contents of that dict. 

For example, something like scores and Q&A pairs I was thinking might look like this:
```python
{
    "score": 4.75,
    "qa_pairs": [("question1", "answer1"), ("question2", "answer2")]
}
```

[the remark this answers is not in the document]
Comment (danmcp) [reply src/instructlab/eval/evaluator.py:15]: If the trade off is between sharing a parent method and having more clarity on what's being returned from the function definition I would vote for the latter.

File: src/instructlab/eval/mtbench.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class MT_Bench_Evaluator(Evaluator):
+    """
+    Child class of an Evaluator for Multi-turn Benchmark (MT-Bench)
+
+    Attributes
+        server  vLLM server endpoint
+    """
+
+    def __init__(self, model, server: str) -> None:
```
Comment (danmcp) [inline src/instructlab/eval/mtbench.py:15]: Should this be server_url

Comment (nathan-weinberg) [reply src/instructlab/eval/mtbench.py:15]: Yeah, I think that's clearer

File: src/instructlab/eval/evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathan-weinberg) [reply src/instructlab/eval/evaluator.py:15]: Yeah this may not be one that makes sense to inherit

File: src/instructlab/eval/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+
+class Evaluator:
+    """
+    Parent class for Evaluators
+
+    Atttributes:
+        model   The model to be evaluated
```
Comment (danmcp) [inline src/instructlab/eval/evaluator.py:9]: Should this be model_path?

Comment (nathan-weinberg) [reply src/instructlab/eval/evaluator.py:9]: Makes sense to me

File: src/instructlab/eval/mmlu.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from .evaluator import Evaluator
+
+
+class MMLU_Evaluator(Evaluator):
+    """
+    Child class of an Evaluator for Massive Multitask Language Understanding (MMLU)
+
+    Attributes:
+        tasks       list of tasks for MMLU to test the model with
+        fewshots    number of examples
```
Comment (danmcp) [inline src/instructlab/eval/mmlu.py:13]: few_shots seems to be the most common syntax

File: src/instructlab/eval/mmlu.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        batchsize   number of GPUs
```
Comment (danmcp) [inline src/instructlab/eval/mmlu.py:14]: batch_size to be consistent

File: src/instructlab/eval/mmlu.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        self.tasks = tasks
+        self.fewshots = fewshots
+        self.batchsize = batchsize
+
+    def run(self) -> dict:
+        individual_scores: dict[str, float] = {}
+        overall_score: float = 0.0
+        payload = {
+            "individual_scores": individual_scores,
+            "overall_score": overall_score,
```
Comment (danmcp) [inline src/instructlab/eval/mmlu.py:30]: Would `return overall_score, individual_scores` be preferable?

Comment (nathan-weinberg) [reply src/instructlab/eval/mmlu.py:30]: I think for @cdoern it would be easier to parse a dict but I'll let him make the final call there

Comment (cdoern) [reply src/instructlab/eval/mmlu.py:30]: I think a single dict makes more sense, less return variability the better

Comment (danmcp) [reply src/instructlab/eval/mmlu.py:30]: Why create a dict just to unpack it?  Plus you can get the type hinting with the tuple.  And wouldn't there be no "parsing" at all with a tuple?

Comment (danmcp) [reply src/instructlab/eval/mmlu.py:30]: > less return variability the better

Is this referring to changes over time not showing up in the definition signature?  The downside though is you are moving your api into the dict keys.  The same breakages will happen on changes, but there is now a naming aspect (of the keys) that needs to be described in the api that won't be evident from the signature.

Comment (cdoern) [reply src/instructlab/eval/mmlu.py:30]: actually, you're right I think I will just be dissecting the dict. The only thing I am wondering is if there's no constant structure, how will I know which values are which?

Comment (danmcp) [reply src/instructlab/eval/mmlu.py:30]: > The only thing I am wondering is if there's no constant structure, how will I know which values are which?

I am not sure I am following.  The call would look something like:

`overall_score, individual_scores = eval.run(...)`

or

`overall_score = eval.run()`

if you only want the first value.  That would be consistent.  If additional return values are ever added to the end, callers would only need to be updated if they want to address the new values.

Comment (nathan-weinberg) [reply src/instructlab/eval/mmlu.py]: I pushed up a possible implementation here. Let me both know your thoughts!

File: src/instructlab/eval/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    """
+    Exception raised when model is not able to be found
+
+    Attributes
+        model   model that is being operated on
+    """
+
+    def __init__(self, model) -> None:
+        super().__init__()
+        self.model = model
```
Comment (danmcp) [inline src/instructlab/eval/exceptions.py:20]: model_path

File: src/instructlab/eval/mmlu.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class MMLU_Evaluator(Evaluator):
+    """
+    Child class of an Evaluator for Massive Multitask Language Understanding (MMLU)
+
+    Attributes:
+        tasks       list of tasks for MMLU to test the model with
+        few_shots    number of examples
+        batch_size   number of GPUs
```
Comment (danmcp) [inline src/instructlab/eval/mmlu.py:14]: Document return values

File: src/instructlab/eval/mtbench.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+# Local
+from .evaluator import Evaluator
+
+
+class MT_Bench_Evaluator(Evaluator):
+    """
+    Child class of an Evaluator for Multi-turn Benchmark (MT-Bench)
+
+    Attributes
+        server_url  vLLM server endpoint
```
Comment (danmcp) [inline src/instructlab/eval/mtbench.py:12]: Document return values

File: src/instructlab/eval/mtbench.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        return overall_score, qa_pairs
+
+
+class PR_Bench_Evaluator(Evaluator):
+    """
+    Child class of an Evaluator for PR-Bench Benchmark (PR-Bench)
+
+    Attributes
+        server_url  vLLM server endpoint
+        questions   questions to be asked
```
Comment (danmcp) [inline src/instructlab/eval/mtbench.py:31]: Document return values

File: src/instructlab/eval/exceptions.py (hunk 7 above)
Comment (nathan-weinberg) [reply src/instructlab/eval/exceptions.py:20]: This one I think would remain `model` since the exception refers to a model and not to a path?

Comment (danmcp) [reply src/instructlab/eval/exceptions.py:20]: Is it the model name?  Whatever it was, it couldn't be found.  I was guessing the only thing that would be available to the calling code is that path.

Comment (nathan-weinberg) [reply src/instructlab/eval/exceptions.py]: I tweaked the implementation to take the path and dynamically determine the model name for the exception message

File: src/instructlab/eval/mtbench.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    Attributes
+        server_url  vLLM server endpoint
+    """
+
+    def __init__(self, model_path, server_url: str) -> None:
+        super().__init__(model_path)
+        self.server_url = server_url
+
+    def run(self) -> tuple:
```
Comment (danmcp) [inline src/instructlab/eval/mtbench.py:19]: Ali needed to change this to gen_answers and judge_answers.

Comment (nathan-weinberg) [reply src/instructlab/eval/mtbench.py:19]: Can we merge this and iterate on it? I'd rather just have the code change subsequently then try to align with something currently in-flight.

Comment (danmcp) [reply src/instructlab/eval/mtbench.py:19]: The other option is to include these commits with the impl commits + any needed modifications.

Comment (nathan-weinberg) [reply src/instructlab/eval/mtbench.py:19]: This is just a skeleton PR - not meant to be all of the implementation as well. I think it makes much more sense to merge this now - it also allows @alinaryan to start from the same point to start the MMLU implementation.

Comment (danmcp) [reply src/instructlab/eval/mtbench.py:19]: Good point about blocking mmlu.  I'm good to merge as is.

**Review by danmcp — APPROVED** [review]

**Pull request merged**